### PR TITLE
Add content-based colorScheme functionality

### LINKED ISSUE #119333: Add support for creating `ColorScheme`s from content
There are [new guidelines on content color](https://m3.material.io/styles/color/dynamic-color/user-generated-color#8af550b9-a19e-4e9f-bb0a-7f611fed5d0f), which is distinct from wallpaper-based color.

image

Already available in Material Color Utilities, we just need to provide API in the framework similar to the existing wallpaper-based color scheme creation.

### PR DESCRIPTION
**Pull request by esouthren**
Fixes https://github.com/flutter/flutter/issues/119333.

Adds `ColorScheme.fromImageProvider()` constructor to Material's color scheme. 

This allows users to pass an `ImageProvider` and receive a Material-based color scheme from the dominant colors of the image.  By using an `ImageProvider`, developers can use local assets, network or memory images. We use the Material Color Utilities library to extract and score the dominant colors of the image. 

The scheme can be in light or dark mode, and additionally, any property of the scheme can be overridden with custom colors if they are preferred over the generated colors.

An `examples/` demo has been added, with a further update to the Material 3 demo app to come. 

The example demo looks as follows: 

https://user-images.githubusercontent.com/15033982/226016022-f7c0ad05-fcec-41b8-bbf5-79e2bfe9bca8.mov

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 1]
```diff
@@ -988,4 +991,194 @@ class ColorScheme with Diagnosticable {
...
+      final ui.PictureRecorder pictureRecorder = ui.PictureRecorder();
+      final Canvas canvas = Canvas(pictureRecorder);
+      paintImage(
+          canvas: canvas,
+          rect: Rect.fromLTRB(0, 0, paintWidth, paintHeight),
+          image: image,
+          filterQuality: FilterQuality.none);
+
+      final ui.Picture picture = pictureRecorder.endRecording();
+      scaledImage =
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/color_scheme.dart:1162]: one line

**Review by HansMuller — APPROVED** [review]
This looks great. Nearly all of the feedback is trivial formatting stuff (mostly just advocating fewer line breaks).

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 2]
```diff
@@ -988,4 +991,194 @@ class ColorScheme with Diagnosticable {
...
+      128,
+      returnInputPixelToClusterPixel: true,
+    );
+    return quantizerResult;
+  }
+
+  // Scale image size down to reduce computation time of color extraction.
+  static Future<ui.Image> _imageProviderToScaled(
+      ImageProvider imageProvider) async {
+    const double maxDimension = 112.0;
```
[the remark this answers is not in the document]
Comment (esouthren) [reply packages/flutter/lib/src/material/color_scheme.dart:1132]: Good point - added.

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 3]
```diff
@@ -988,4 +991,194 @@ class ColorScheme with Diagnosticable {
...
+    final QuantizerResult quantizerResult = await QuantizerCelebi().quantize(
+      imageBytes!.buffer.asUint32List(),
+      128,
+      returnInputPixelToClusterPixel: true,
+    );
+    return quantizerResult;
+  }
+
+  // Scale image size down to reduce computation time of color extraction.
+  static Future<ui.Image> _imageProviderToScaled(
```
[the remark this answers is not in the document]
Comment (esouthren) [reply packages/flutter/lib/src/material/color_scheme.dart:1130]: Done

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 4]
```diff
@@ -988,4 +991,194 @@ class ColorScheme with Diagnosticable {
...
+      if (rescale) {
+        paintWidth =
+            (width > height) ? maxDimension : (maxDimension / height) * width;
+        paintHeight =
+            (height > width) ? maxDimension : (maxDimension / width) * height;
+      }
+      final ui.PictureRecorder pictureRecorder = ui.PictureRecorder();
+      final Canvas canvas = Canvas(pictureRecorder);
+      paintImage(
+          canvas: canvas,
```
[the remark this answers is not in the document]
Comment (esouthren) [reply packages/flutter/lib/src/material/color_scheme.dart:1156]: Done

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 5]
```diff
@@ -988,4 +991,194 @@ class ColorScheme with Diagnosticable {
...
+          rect: Rect.fromLTRB(0, 0, paintWidth, paintHeight),
+          image: image,
+          filterQuality: FilterQuality.none);
+
+      final ui.Picture picture = pictureRecorder.endRecording();
+      scaledImage =
+          await picture.toImage(paintWidth.toInt(), paintHeight.toInt());
+      imageCompleter.complete(info.image);
+    }, onError: (_, __) {
```
[the remark this answers is not in the document]
Comment (esouthren) [reply packages/flutter/lib/src/material/color_scheme.dart:1165]: Done, thanks

**Pull request merged**